Support iframes (nested browsing contexts) in selection event handling (stale version)

acusti opened the pull request:
As mentioned in https://github.com/facebook/react/pull/7866#issuecomment-252313617, this PR takes the approach of checking all nested browsing contexts and saving and restoring their selections regardless of whether or not they are the primary focused element on the page. It is a more complete alternative to the solution in #7866, and like that PR, fixes #427. It is also, like the original, dependent on facebook/fbjs#188, but that PR has been merged into master, so as soon as a new fbjs version is released, this will be ready to go.

acusti commented:
You can see this version of the solution in practice with an actual use case in this demo: http://codepen.io/acusti/pen/yajJbZ?editors=0010

The essential point is that I no longer need to manually manage focus in the app code and instead can rely on the ReactInputSelection functionality to take care of preserving visible “focus” (i.e. visible selections).

acusti commented:
If anyone wants to play around with this, I pushed up a copy of the built react and react-dom scripts:

```
https://rawgit.com/Brandcast/react/nested-browsing-contexts-2-built/build/react.js
https://rawgit.com/Brandcast/react/nested-browsing-contexts-2-built/build/react-dom.js
```

acusti commented on src/renderers/dom/client/__tests__/ReactInputSelection-test.js:
```diff
@@ -0,0 +1,227 @@
+
+    it('gets and restores selection for inputs in an iframe that get remounted', () => {
+      var iframe = document.createElement('iframe');
+      iframe.setAttribute('tabIndex', 0);
```
Setting tabIndex on the iframe element here and then invoking `iframe.focus()` below at line 202 was only necessary because jsdom didn’t handle setting `activeElement` in the parent correctly. This is fixed in the jsdom v9.7.0, so if we updated jest’s [jsdom dependency](https://github.com/facebook/jest/blob/master/package.json#L23), I could remove the hacky workaround.

acusti replied:
Updating the version of jsdom used by jest [has landed](https://github.com/facebook/jest/pull/1944) (includes a fix of a major perf regression, so it’ll be fun to see how that affects test running times), so when a new version is released, should be able to remove the hacky workarounds in this test.

acusti commented:
If anyone wants npm-installable versions of this PR to experiment with, you can use these in your package.json (requires the latest `fbjs` changes in master which have not yet been released, which is why that dependency also needs to be included):

```json
    "fbjs": "Brandcast/fbjs-built#97b8e54",
    "react": "Brandcast/react#3b54263",
```

acusti commented:
@spicyj I’ve put together a Draft.js demo of this PR along with the Draft.js PR to illustrate the utility of making React work across nested browsing contexts. The PR is facebook/draft-js#765, and the demo is here: http://codepen.io/acusti/pen/RGEJZE?editors=0011

In that demo, I used built versions of React and Draft.js which include all the changes for working across nested browsing contexts. Those built versions are usable in node-land or the browser:

**In node**

```js
    "draft-js": "Brandcast/draft-js#f9affa3",
    "fbjs": "Brandcast/fbjs-built#97b8e54",
    "react": "Brandcast/react#9a81d1a",
```

**In the browser**

```html
    <script src="https://rawgit.com/Brandcast/react/nested-browsing-contexts-2-built/build/react.js"></script>
    <script src="https://rawgit.com/Brandcast/react/nested-browsing-contexts-2-built/build/react-dom.js"></script>
    <script src="https://cdn.jsdelivr.net/immutable.js/latest/immutable.min.js"></script>
    <script src="https://cdn.jsdelivr.net/es6.shim/0.35.1/es6-shim.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-core/5.8.34/browser.js"></script>
    <script src="https://rawgit.com/Brandcast/draft-js/iframe-compatibility-built/dist/Draft.js"></script>
```

I also rebased my branch (for this PR to React) to remove all upstream conflicts.

acusti commented:
@spicyj I’ve put together a Draft.js demo of this PR along with the Draft.js PR to illustrate the utility of making React work across nested browsing contexts. The PR is facebook/draft-js#765, and the demo is here: http://codepen.io/acusti/pen/RGEJZE?editors=0011

In that demo, I used built versions of React and Draft.js which include all the changes for working across nested browsing contexts. Those built versions are usable in node-land or the browser:

**In node**

```js
    "draft-js": "Brandcast/draft-js#f9affa3",
    "fbjs": "Brandcast/fbjs-built#97b8e54",
    "react": "Brandcast/react#9a81d1a",
```

**In the browser**

```html
    <script src="https://rawgit.com/Brandcast/react/nested-browsing-contexts-2-built/build/react.js"></script>
    <script src="https://rawgit.com/Brandcast/react/nested-browsing-contexts-2-built/build/react-dom.js"></script>
    <script src="https://cdn.jsdelivr.net/immutable.js/latest/immutable.min.js"></script>
    <script src="https://cdn.jsdelivr.net/es6.shim/0.35.1/es6-shim.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-core/5.8.34/browser.js"></script>
    <script src="https://rawgit.com/Brandcast/draft-js/iframe-compatibility-built/dist/Draft.js"></script>
```

I also rebased my branch (for this PR to React) to remove all upstream conflicts.

spicyj commented:
Sorry, I haven't had time to look at this. The intent is good though and I would like to get it merged.

acusti commented:
Just updated the PR after rebasing with master. There was a conflict in `ReactInputSelection` as a result of the new code added to preserve scroll position when restoring selection and focusing an element. I incorporated that new functionality into these changes.

I also verified that I could remove the hacky workaround in the new ReactInputSelection tests thanks to jsdom being updated to have more accurate iframe focus behavior, so the tests are a bit cleaner now. And `npm test` is all passing.

acusti commented:
Thought I should mention that I’ve also updated my built versions of everything, including draft-js (v0.9.0), react (v15.4.1), and react-dom (v15.4.1). This is particularly important because the previous built versions were based on the React master branch 26 days ago (i.e. 16.0.0-alpha, or fiber), whereas these built versions are built from 15.4.1. To try out the PRs:

### npm / yarn

```json
    "draft-js": "brandcast/draft-js-built#333bf90",
    "fbjs": "brandcast/fbjs-built#3961252",
    "react": "brandcast/react-built#ab31c46",
    "react-dom": "brandcast/react-dom-built#6cd1db3",
```

### in the browser

```html
    <script src="https://rawgit.com/brandcast/react-built/15-stable/dist/react.js"></script>
    <script src="https://rawgit.com/brandcast/react-dom-built/15-stable/dist/react-dom.js"></script>
    <script src="https://cdn.jsdelivr.net/immutable.js/latest/immutable.min.js"></script>
    <script src="https://cdn.jsdelivr.net/es6.shim/0.35.1/es6-shim.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-core/5.8.34/browser.js"></script>
    <script src="https://rawgit.com/brandcast/draft-js-built/master/dist/Draft.js"></script>
```

aleksey-shmatov commented:
There is minor issue in Safari. I have one iframe with my content and one iframe for Google Maps. My website accessed with "http" and Google Maps hosted with "https". In the getElementsWithSelections following code does not throw error: 
`try {doc = win.document;} catch (e) {return acc;}` 
"doc" will be undefined after this code. This causes "undefined is not an object" exception and breaks functionality.

acusti commented:
Interesting. I suppose just adding an additional early return if doc is falsy would handle that. Thanks for the feedback!

aleksey-shmatov commented:
Thanks for adding this functionality, hope it get merged into master soon(together with draft-js). I have been working with this for about a week now and the only issue I encountered was that one for Safari.

philipp-spiess commented:
@spicyj What's the status here? We'd need that behaviour but we can't afford to maintain a fork of React. Do you still have plans to merge this? Your last update is unfortunately already a few months old.

acusti commented:
@sebmarkbage To clarify, did you remove the `GH Review: review-needed` label because someone did a review?

spicyj commented:
@acusti No, we have switched to using the github review feature for all PRs.

acusti commented:
I tried to rebase this branch from master and ran into a world of hurt, so I created a fresh branch off latest master cherry-picked over all the commits, and made a new PR: #9184

@aleksey-shmatov, I added a guard for the Safari bug you found in f8b202d, as well as dealing with a Firefox issue I found in my own testing. Thanks for trying it out and reporting back! I’ll hoping to make new built versions of everything off of these commits applied to 15-stable and update the npm / yarn instructions on using those soon. When I do, I will update both this now-stale PR and the #9184 with the updated dependency strings.

acusti commented:
For anyone who has been following along with this effort, I’m closing this in favor of a new PR: #11410

If you want to use the changes in the latest PR, which is based off of react 16, in your own project, you can do so by updating the `react` and `react-dom` dependencies in your `package.json` to:

```json
    "react": "brandcast/react-built#7c8fc6b",
    "react-dom": "brandcast/react-dom-built#017534a",
```

If you’re also using draft.js and want the full latest and greatest cross-iframe experience, use:

```json
    "draft-js": "brandcast/draft-js-built#003f24b",
```
